### ISSUE TITLE
Question about Top-p implementation

### ISSUE THREAD

**Issue by Klaw807** [tsinghua-ideal/Twilight#6]
Thanks for releasing the code and the paper! I’m very interested in the Twilight / hierarchical top-p pruning idea.

I checked the current efficiency benchmark in the README:

```bash
cd benchmark/efficiency
python3 bench_gemv.py
```

From my understanding, the current implementation seems mainly to benchmark a sparse GEMV / fixed-budget path, which looks closer to a top-k-style implementation. I could not find the full top-p version described in the paper, especially the adaptive pruning part where different heads may keep different numbers of tokens.

Could you please let me know whether the real top-p implementation is included somewhere in the repo? If not, I’ll appreciate it a lot if you could provide the top-p version or the benchmark scripts used for the Twilight efficiency results.

**Comment by SiriusNEO** [comment]
Hi @Klaw807 ! Sorry for the late reply. The top-p kernel was already provided (using the FlashInfer algorithm, but only the mask version). Now I updated more efficiency evaluation scripts and kernels, including:
- The indices version of top-p kernel (Previously it's mask version, which is not convient to be integrated into full computational flow).
- Top-p benchmark: `benchmark/efficiency/bench_top_p.py`
- Twilight operator breakdown: `benchmark/efficiency/bench_breakdown.py`, which also provides a full version of the Twilight's attention layer.
- Per-head ragged / GQA attention benchmark: `benchmark/efficiency/bench_gqa.py`

PS: the public benchmark uses synthetic logits/cache tensors, so the exact pruning budget and speedups may differ from the paper’s trace-based measurements.

Hoping this can help!
